File: app/lib/models/gps_boundary.dart
import 'dart:math' as math;
import 'package:json_annotation/json_annotation.dart';

part 'gps_boundary.g.dart';

@JsonSerializable()
class GPSBoundary {
  final String id;
  final String? clientId;
  final String? siteId;
  final String name;
  final double centerLatitude;
  final double centerLongitude;
  final double radiusMeters;
  final int priority;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool isActive;

  GPSBoundary({
    required this.id,
    this.clientId,
    this.siteId,
    required this.name,
    required this.centerLatitude,
    required this.centerLongitude,
    required this.radiusMeters,
    required this.priority,
    required this.createdAt,
    required this.updatedAt,
    this.isActive = true,
  });

  factory GPSBoundary.fromJson(Map<String, dynamic> json) => _$GPSBoundaryFromJson(json);
  Map<String, dynamic> toJson() => _$GPSBoundaryToJson(this);

  GPSBoundary copyWith({
    String? id,
    String? clientId,
    String? siteId,
    String? name,
    double? centerLatitude,
    double? centerLongitude,
    double? radiusMeters,
    int? priority,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isActive,
  }) {
    return GPSBoundary(
      id: id ?? this.id,
      clientId: clientId ?? this.clientId,
      siteId: siteId ?? this.siteId,
      name: name ?? this.name,
      centerLatitude: centerLatitude ?? this.centerLatitude,
      centerLongitude: centerLongitude ?? this.centerLongitude,
      radiusMeters: radiusMeters ?? this.radiusMeters,
      priority: priority ?? this.priority,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isActive: isActive ?? this.isActive,
    );
  }

  bool isValidCoordinates() {
    return centerLatitude >= -90 &&
        centerLatitude <= 90 &&
        centerLongitude >= -180 &&
        centerLongitude <= 180;
  }

  bool isValidRadius() {
    return radiusMeters > 0 && radiusMeters <= 10000;
  }

  bool containsLocation(double latitude, double longitude) {
    final distance = _calculateDistance(
      centerLatitude,
      centerLongitude,
      latitude,
      longitude,
    );
    return distance <= radiusMeters;
  }

  double distanceToLocation(double latitude, double longitude) {
    return _calculateDistance(
      centerLatitude,
      centerLongitude,
      latitude,
      longitude,
    );
  }

  double _calculateDistance(double lat1, double lon1, double lat2, double lon2) {
    const earthRadius = 6371000.0;
    final dLat = _toRadians(lat2 - lat1);
    final dLon = _toRadians(lon2 - lon1);
    final a = math.sin(dLat / 2) * math.sin(dLat / 2) +
        math.cos(_toRadians(lat1)) *
            math.cos(_toRadians(lat2)) *
            math.sin(dLon / 2) *
            math.sin(dLon / 2);
    final c = 2 * math.asin(math.sqrt(a));
    return earthRadius * c;
  }

  double _toRadians(double degrees) => degrees * (3.14159265359 / 180.0);

  double get confidence {
    if (radiusMeters <= 100) return 1.0;
    if (radiusMeters <= 500) return 0.9;
    if (radiusMeters <= 1000) return 0.8;
    if (radiusMeters <= 5000) return 0.6;
    return 0.4;
  }

  bool overlapsWith(GPSBoundary other) {
    final distance = _calculateDistance(
      centerLatitude,
      centerLongitude,
      other.centerLatitude,
      other.centerLongitude,
    );
    return distance < (radiusMeters + other.radiusMeters);
  }

  int compareTo(GPSBoundary other) {
    if (priority != other.priority) {
      return other.priority.compareTo(priority);
    }
    return radiusMeters.compareTo(other.radiusMeters);
  }
}

File: api/src/models/gps_boundary.js
const { DataTypes } = require('sequelize');
const sequelize = require('../database/connection');

const GPSBoundary = sequelize.define('GPSBoundary', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  clientId: {
    type: DataTypes.UUID,
    allowNull: true,
    references: {
      model: 'Clients',
      key: 'id',
    },
  },
  siteId: {
    type: DataTypes.UUID,
    allowNull: true,
    references: {
      model: 'Sites',
      key: 'id',
    },
  },
  name: {
    type: DataTypes.STRING(100),
    allowNull: false,
    validate: {
      notEmpty: true,
      len: [1, 100],
    },
  },
  centerLatitude: {
    type: DataTypes.DECIMAL(10, 8),
    allowNull: false,
    validate: {
      min: -90,
      max: 90,
    },
  },
  centerLongitude: {
    type: DataTypes.DECIMAL(11, 8),
    allowNull: false,
    validate: {
      min: -180,
      max: 180,
    },
  },
  radiusMeters: {
    type: DataTypes.DECIMAL(10, 2),
    allowNull: false,
    validate: {
      min: 1,
      max: 10000,
    },
  },
  priority: {
    type: DataTypes.INTEGER,
    allowNull: false,
    defaultValue: 1,
    validate: {
      min: 1,
    },
  },
  createdAt: {
    type: DataTypes.DATE,
    allowNull: false,
  },
  updatedAt: {
    type: DataTypes.DATE,
    allowNull: false,
  },
  isActive: {
    type: DataTypes.BOOLEAN,
    defaultValue: true,
  },
}, {
  tableName: 'gps_boundaries',
  timestamps: true,
  indexes: [
    {
      fields: ['clientId'],
    },
    {
      fields: ['siteId'],
    },
    {
      fields: ['priority'],
    },
    {
      fields: ['isActive'],
    },
  ],
});

GPSBoundary.prototype.toJSON = function() {
  const values = { ...this.get() };
  return values;
};

GPSBoundary.prototype.containsLocation = function(latitude, longitude) {
  const distance = this.calculateDistance(
    this.centerLatitude,
    this.centerLongitude,
    latitude,
    longitude
  );
  return distance <= this.radiusMeters;
};

GPSBoundary.prototype.distanceToLocation = function(latitude, longitude) {
  return this.calculateDistance(
    this.centerLatitude,
    this.centerLongitude,
    latitude,
    longitude
  );
};

GPSBoundary.prototype.calculateDistance = function(lat1, lon1, lat2, lon2) {
  const earthRadius = 6371000;
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLon = (lon2 - lon1) * Math.PI / 180;
  const a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(lat1 * Math.PI / 180) *
    Math.cos(lat2 * Math.PI / 180) *
    Math.sin(dLon / 2) *
    Math.sin(dLon / 2);
  const c = 2 * Math.asin(Math.sqrt(a));
  return earthRadius * c;
};

GPSBoundary.prototype.getConfidence = function() {
  const radius = parseFloat(this.radiusMeters);
  if (radius <= 100) return 1.0;
  if (radius <= 500) return 0.9;
  if (radius <= 1000) return 0.8;
  if (radius <= 5000) return 0.6;
  return 0.4;
};

GPSBoundary.prototype.overlapsWith = function(other) {
  const distance = this.calculateDistance(
    this.centerLatitude,
    this.centerLongitude,
    other.centerLatitude,
    other.centerLongitude
  );
  return distance < (parseFloat(this.radiusMeters) + parseFloat(other.radiusMeters));
};

GPSBoundary.findByLocation = async function(latitude, longitude, companyId = null) {
  const allBoundaries = await this.findAll({
    where: {
      isActive: true,
      ...(companyId && { clientId: companyId }),
    },
  });

  const matches = [];
  for (const boundary of allBoundaries) {
    if (boundary.containsLocation(latitude, longitude)) {
      matches.push({
        boundary,
        distance: boundary.distanceToLocation(latitude, longitude),
        confidence: boundary.getConfidence(),
      });
    }
  }

  matches.sort((a, b) => {
    if (a.boundary.priority !== b.boundary.priority) {
      return b.boundary.priority - a.boundary.priority;
    }
    return a.distance - b.distance;
  });

  return matches;
};

GPSBoundary.findByClient = function(clientId) {
  return this.findAll({
    where: {
      clientId,
      isActive: true,
    },
    order: [['priority', 'DESC'], ['name', 'ASC']],
  });
};

GPSBoundary.findBySite = function(siteId) {
  return this.findAll({
    where: {
      siteId,
      isActive: true,
    },
    order: [['priority', 'DESC'], ['name', 'ASC']],
  });
};

GPSBoundary.associate = function(models) {
  GPSBoundary.belongsTo(models.Client, {
    foreignKey: 'clientId',
    as: 'client',
  });
  GPSBoundary.belongsTo(models.Site, {
    foreignKey: 'siteId',
    as: 'site',
  });
};

module.exports = GPSBoundary;